libraries/retrying_amqp_connection/include/eosio/retrying_amqp_connection/retrying_amqp_connection.hpp

```diff
@@ -0,0 +1,268 @@
+     _executor(executor), _resolver(executor), _sock(executor), _timer(executor), _address(address),
+     _ready_callback(ready), _failed_callback(failed), _logger(logger) {
+
+      FC_ASSERT(!_address.secure(), "Only amqp:// URIs are supported for AMQP addresses (${a})", ("a", _address));
+      start_connection();
+   }
+
+   const AMQP::Address& address() const {
+      return _address;
+   }
+
+protected:
+   void onReady(AMQP::Connection* connection) override {
+      fc_ilog(_logger, "AMQP connection to ${s} is fully operational", ("s", _address));
+
+      _ready_callback(connection);
+      _indicated_ready = true;
+   }
+
+   void onData(AMQP::Connection* connection, const char* data, size_t size) override {
+      if(!_sock.is_open())
+         return;
+      _state->outgoing_queue.emplace_back(data, data+size);
+      send_some();
+   }
+
+   void onError(AMQP::Connection* connection, const char* message) override {
+      fc_elog(_logger, "AMQP connection to ${s} suffered an error; will retry shortly: ${m}", ("s", _address)("m", message));
+      schedule_retry();
+   }
+
+   void onClosed(AMQP::Connection *connection) override {
+      fc_wlog(_logger, "AMQP connection to ${s} closed AMQP connection", ("s", _address));
+      schedule_retry();
+   }
+
+private:
+  void start_connection() {
+       _resolver.async_resolve(_address.hostname(), std::to_string(_address.port()), boost::asio::bind_executor(_executor, [this](const auto ec, const auto endpoints) {
+         if(ec) {
+            if(ec != boost::asio::error::operation_aborted) {
+               fc_wlog(_logger, "Failed resolving AMQP server ${s}; will retry shortly: ${m}", ("s", _address)("m", ec.message()));
+               schedule_retry();
+            }
+            return;
+         }
+         //AMQP::Connection's dtor will attempt to send a last gasp message. Resetting state here is a little easier to prove
+         // as being safe as it requires pumping the event loop once vs placing the state reset directly in schedule_retry()
+         _state.emplace();
+         boost::asio::async_connect(_sock, endpoints, boost::asio::bind_executor(_executor, [this](const auto ec, const auto endpoint) {
+            if(ec) {
+               if(ec != boost::asio::error::operation_aborted) {
+                  fc_wlog(_logger, "Failed connecting AMQP server ${s}; will retry shortly: ${m}", ("s", _address)("m", ec.message()));
+                  schedule_retry();
+               }
+               return;
+            }
+            fc_ilog(_logger, "TCP connection to AMQP server at ${s} is up", ("s", _address));
+            receive_some();
+            _state->amqp_connection.emplace(this, _address.login(), _address.vhost());
+         }));
+      }));
+   }
+
+   void schedule_retry() {
+      if(_indicated_ready)
+         _failed_callback();
+      _indicated_ready = false;
+      _sock.close();
+      _resolver.cancel();
```

Comment:
It looks like `schedule_retry` is only called from callbacks bound via `boost::asio::bind_executor(_executor...`.
To be safe, I think `_resolver.async_resolve` should be called via `_executer.post()` and then everything will be on the executor strand. The class comment says to use a strand where appropriate, which I read to mean give the class a strand everywhere where you are not using `app().get_io_service()` and no other threads.

Reply:
I'm think Steven's comment stems from the implicit vs explicit strand discussion we had once before: even if you're running an io_context in a single thread (which constitutes an implicit strand according asio docs) there appears to be no guarantee that is spelled out anywhere that an implicit strand services its queue in FIFO ordering. Compare to an explicit strand (by construction of a strand object) that does clearly state in the docs it enforces FIFO ordering.

Originally I felt that was a rather pessimistic read of the docs however there is clear evidence in the Networking TS that a stand object is required for FIFO ordering. In particular see
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1442r0.html#_lack_of_ordering_guarantees_of_a_single_threaded_code_io_context_code
which was reviewed some point here
```
https://github.com/cplusplus/papers/issues/211#issuecomment-510784724
(in a code block so I don't cause issue github auto-linkage)
```
but I can't find evidence of what happened after this.

This code was written the way it was (templated Executor) because I didn't want someone to incur the cost of using a strand if they didn't need it. But if something in the code requires FIFO ordering then it seems a strand is required.